new exercise alphametics

abo64 opened the pull request:
of course feel free to change the place of the exercise in `config.json`. First I thought it makes sense to have it after `sgf-parsing` but then I saw its difficulty of 9 and reconsidered.
I kept the tests for multiplication and exponentiation from a former version of `canconical-data.json` as they make the exercise more interesting.
what annoys me is the more than 30 times worse performance than my same (I think) Scala implementation. would be very much interested to know the reason why.

rbasso commented on exercises/alphametics/test/Tests.hs:
```diff
@@ -0,0 +1,76 @@
+data Case = Case { description  ::        String
+                 , puzzle       ::        String
+                 , expected     ::        Maybe[(Char, Int)]
+                 }
```
##### Cosmetic suggestion

A lot of unneeded spaces. One missing space after `Maybe`.

I think that this would be more readable:
```haskell
data Case = Case { description :: String
                 , puzzle      :: String
                 , expected    :: Maybe [(Char, Int)]
                 }
```

rbasso commented on exercises/alphametics/test/Tests.hs:
```diff
@@ -0,0 +1,76 @@
+cases = [ Case { description  = "puzzle with three letters"
+               , puzzle       = "I + BB == ILL"
+               , expected     = Just [('I', 1), ('B', 9), ('L', 0)]
+               }
```
##### Cosmetic suggestion

I would remove one space before each `=` here, aligning everything with `description =`

rbasso commented on exercises/alphametics/test/Tests.hs:
```diff
@@ -0,0 +1,76 @@
+               }
+        , Case { description  = "puzzle with multiplication"
+               , puzzle       = "IF * DR == DORI"
+               , expected     = Just [('I', 8), ('F', 2), ('D', 3), ('R', 9), ('O', 1)]
```
These two last exercises with `*` and `^` seem to make the exercise significantly harder, forcing people to learn a lot about parsing before even starting to solve the actual puzzle.

Most of the users [give up on the track](http://www.exercism.io/stats/haskell) before reaching the final exercises, so I think we should focus on making the exercises more pedagogical to the less advanced students

I found that this was already discussed in https://github.com/exercism/x-common/issues/395, so I think that in this case we should follow `x-common` and remove them.

What do you think about it, @petertseng ?

abo64 replied:
thx @rbasso  I will try to implement all your helpful suggestions.
I agree that the last two tests make the whole exercise harder, and we could leave them out. Or perhaps make them somehow optional by commenting them out/marking them as pending so that each student can decide for himself?
Main problem is of course the poor performance. And I still wonder why. As mentioned above it works fine in Scala. Only explanation so far I could come up with is that something causes most or all `permutations` to become evaluated. In Scala `permutations` returns an Iterator, so this might make sure that the whole computation remains lazy? I will try again to find the reason - this is an interesting problem in itself. And of course - any ideas/help appreciated.

abo64 replied:
Fine then, I will remove them.
That means I will go back to the drawing board for the example solution as the current one indeed is brute force using `permutations` (and has this lousy performance in comparison to Scala; although I still intend to find the reason why, but later then).

petertseng replied:
It could still be the case that many students will still submit the simple brute force solution anyway. It's the first thing that comes to mind. Different people want to see different things from Exercism. So limiting to addition problems opens the door to doing something more considered, for those who want to do that. The decreased difficult does help a bit too, I think.

I realize that another way to expanding this problem is indeed to keep the operations other than addition. There are various directions to take this problem. I chose one way (and it happens to be the way that x-common chose too), but maybe others like a different way.

As for what we can do about this long runtime, maybe we can get some ideas from other tracks. https://github.com/exercism/xruby/pull/455/files#diff-8ec059eaf4c70572282d3173a6325c3aR53 the Ruby track chose to comment out the cases that had many letters since they do take a long time to solve that way. Recently as of https://github.com/exercism/xruby/pull/469 the Ruby example is fast enough to solve the equations with many letters, but they still have a comment telling students it's OK not to do them: https://github.com/exercism/xruby/pull/469/files#diff-8ec059eaf4c70572282d3173a6325c3aR53 . If we decide here that the same ideas are appropriate, we could apply them.

petertseng replied:
> I'll admit I'm not 100% sure either, since there's a lot going on.

But I can surely explain what I would do, if you are interested

abo64 replied:
yeah, starting from the rightmost digit/column, doing the sum and then accumulating restrictions going left.
I thought that this might even work for multiplication and exponentiation.

petertseng replied:
I agree it can work with multiplication - we would have to consider more columns at a time, since in:

```
 AB
xCD
---
EFG

both A*D and B*C contribute to F.

But once that detail is ironed out, it would still be doable. I have not tested it out myself to make sure, but seems that way, anyway.

Exponentiation would thus also be doable if we just treat it as repeated multiplication, assuming there are no variable exponents.

abo64 commented on exercises/alphametics/src/Alphametics.hs:
```diff
@@ -0,0 +1,4 @@
+module Alphametics (solve) where
+
+solve :: String -> Maybe [(Char, Int)]
+solve puzzle = undefined
```
I thought so, too. One can implement a solution using a `Map`, say, and then easily transform to this basic type.
About including `puzzle` I considered
```haskell
type Puzzle = String

solve :: Puzzle -> Maybe [(Char, Int)]
solve = undefined
```
but decided against it as this coding style with `type`s might not be to everybody's liking.

abo64 commented on exercises/alphametics/package.yaml:
```diff
@@ -0,0 +1,20 @@
+  dependencies:
+  # - foo       # List here the packages you
+  # - bar       # want to use in your solution.
+    - parsec
```
`word-count` looks good. at least some `HINTS.md` might be a good compromise.

abo64 commented:
I have implemented this column-wise partial solution algorithm. So far it only works for addition and only up to eight letter words. Thus I moved the ten letter test as well as the tests for multiplication and exponentiation to `HINTS.md`.

rbasso commented on exercises/alphametics/HINTS.md:
````diff
@@ -0,0 +1,25 @@
+               , puzzle       = "PI * R ^ 2 == AREA"
+               , expected     = Just [('P', 9), ('I', 6), ('R', 7), ('A', 4), ('E', 0)]
+               }
+```
````
I feel that these cases pollute the `HINTS.md` file. Also, I don't believe anyone will copy/paste this in `test/Test.hs`.

I prefer them commented out in the test suite, but I will respect your decision if you prefer them here, @abo64.

abo64 commented:
I moved the tests back to `Tests.hs`. Shall we then keep `HINTS.md`?
And I am afraid I have done some mistake while squashing? :-(

rbasso commented:
You can use `git rebase -i master` and delete the line with the unwanted commit to fix that.

abo64 commented:
If it is too messed up: I could create a new PR with just the final changes?

rbasso commented:
What a mess, man! :smile: 

Don't worry!

I'm not good at *git*, but I think that a I have a solution for this problem:

- Type `git rebase --abort`, just to be sure that there is nothing ongoin.
- Check that you are in the branch named `exercise-alphametics`.
- Type `git rebase -i 82ae06`
- Remove lines until it matches **exactly** what is below:

```git
pick 7c98beb new exercise alphametics                                                                           
squash 5d75fb0 fix hlint complaint                                              
squash 39cc180 example with column-wise addition algorithm, removed problematic tests
squash dcb567e moved additional tests back to Tests.hs                          
``` 
- Edit the commit message to something like:

```
alphametics: add new exercise
```

- Push the changes to your fork: `git push -f`

abo64 commented:
thanks, man! :-)

rbasso commented:
Now you probably will want to fix your `master` branch. I think that something like this will work:

```
$ git checkout master
$ git reset --hard upstream/master
$ git pull upstream master
$ git push -f origin master

abo64 commented:
done.

rbasso commented:
Thanks, @abo64! I know that this one gave you a lot of work! :+1:
